File: app/src/main/java/com/vikingsen/cheesedemo/ux/cheeselist/CheeseListPresenter.java
package com.vikingsen.cheesedemo.ux.cheeselist;


import android.support.annotation.Nullable;

import com.vikingsen.cheesedemo.model.data.cheese.CheeseRepository;
import com.vikingsen.cheesedemo.util.SchedulerProvider;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;

class CheeseListPresenter {

    private final CheeseListContract.View view;
    private final CheeseRepository cheeseRepository;
    private final SchedulerProvider schedulerProvider;

    @Nullable
    private Disposable cheeseDisposable = null;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Inject
    CheeseListPresenter(CheeseListContract.View view,
                        CheeseRepository cheeseRepository,
                        SchedulerProvider schedulerProvider) {
        this.view = view;
        this.cheeseRepository = cheeseRepository;
        this.schedulerProvider = schedulerProvider;
    }

    void start() {
        loadCheeses(false);
    }

    void stop() {
        compositeDisposable.clear();
    }

    void loadCheeses(boolean forceRefresh) {
        view.showLoading(true);
        if (cheeseDisposable != null) {
            cheeseDisposable.dispose();
            compositeDisposable.remove(cheeseDisposable);
        }
        cheeseDisposable = cheeseRepository.getCheeses(forceRefresh)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(cheeses -> {
                            view.showCheeses(cheeses);
                            view.showLoading(false);
                        },
                        throwable -> {
                            Timber.e(throwable, "Error loading cheeses");
                            view.showError();
                            view.showLoading(false);
                        });
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/ux/cheeselist/CheeseListPresenter.kt
package com.vikingsen.cheesedemo.ux.cheeselist


import com.vikingsen.cheesedemo.model.data.cheese.CheeseRepository
import com.vikingsen.cheesedemo.util.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import timber.log.Timber
import javax.inject.Inject

class CheeseListPresenter
@Inject constructor(private val view: CheeseListContract.View,
                    private val cheeseRepository: CheeseRepository,
                    private val schedulerProvider: SchedulerProvider) {

    private var cheeseDisposable: Disposable? = null
    private val compositeDisposable = CompositeDisposable()

    fun start() {
        loadCheeses(false)
    }

    fun stop() {
        compositeDisposable.clear()
    }

    fun loadCheeses(forceRefresh: Boolean) {
        view.showLoading(true)
        cheeseDisposable?.let {
            it.dispose()
            compositeDisposable.remove(it)
        }
        cheeseDisposable = cheeseRepository.getCheeses(forceRefresh)
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        {
                            view.showCheeses(it)
                            view.showLoading(false)
                        },
                        {
                            Timber.e(it, "Error loading cheeses")
                            view.showError()
                            view.showLoading(false)
                        }
                )
    }
}
